### PR TITLE
Check if the localpart is reserved for guests earlier in the registration flow

### LINKED ISSUE #7576: When using numeric only id, registration fails but after checking captcha, mail, ...
This error https://github.com/matrix-org/synapse/blob/b2b86990705de8a099093ec141ad83e09f182034/synapse/handlers/register.py#L177

You get this error very late in the register flow, a bit annoying to know it only after having completed all the register steps

Can it be checked like for 'User ID already taken.'?

image

### PR DESCRIPTION
**Pull request by babolivier**
This is so the user is warned about the username not being valid as soon as possible, rather than only once they've finished UIA.

Fixes #7576

### CODE REVIEW

File: synapse/handlers/register.py [hunk 1]
```diff
@@ -128,6 +128,15 @@ def check_username(self, localpart, guest_access_token=None, assigned_user_id=No
                     errcode=Codes.FORBIDDEN,
                 )
 
+        if guest_access_token is not None:
```
Comment (clokep) [inline synapse/handlers/register.py:131]: I think this logic is reversed, the removed logic is:

```python
was_guess = guest_access_token is not None
if not was_guess:
    ...
```

So I think this should be `if guest_access_token is None`.

Comment (babolivier) [reply synapse/handlers/register.py:131]: Ugh, yes, thanks

Comment (babolivier) [reply synapse/handlers/register.py]: That explains why the sytests were all failing fwiw.

File: changelog.d/7625.misc [hunk 2]
```diff
@@ -0,0 +1 @@
+Check if the localpart is reserved for guests users earlier in the registration flow, as well as when responding to requests on `/register/available`.
```
Comment (clokep) [inline changelog.d/7625.misc:1]:
```suggestion
Check if the localpart of a Matrix ID is reserved for guest users earlier in the registration flow, as well as when responding to requests to `/register/available`.
```

File: synapse/handlers/register.py [hunk 3]
```diff
@@ -128,6 +128,15 @@ def check_username(self, localpart, guest_access_token=None, assigned_user_id=No
                     errcode=Codes.FORBIDDEN,
                 )
 
+        if guest_access_token is None:
+            try:
+                int(localpart)
+                raise SynapseError(
+                    400, "Numeric user IDs are reserved for guest users."
```
Comment (bmarty) [inline synapse/handlers/register.py:135]: Could be nice to send errcode `M_INVALID_USERNAME`, for the moment, the client receives `M_UNKNOWN`:

```json
{"errcode":"M_UNKNOWN","error":"Numeric user IDs are reserved for guest users."}
```

**Pull request merged**